### PR TITLE
Remove Timestamp from Field

### LINKED ISSUE #1384: Create FieldState without timestamp
### Description of the task

For certain applications like initializing the state of a physics simulation, it is useful to have a representation of the field that does not contain timestamp data because it does not make sense in this case.

We should split our `Field` class slightly to support non-timestamped data as well.

### Acceptance criteria

- [ ] Split `FIeld` to also handle non-timestamped data

### PR DESCRIPTION
**Pull request by MathewMacDougall**
### Description

Split the `Field` class into `FieldState` (which contains no timestamp data), and `Field`, which is a timestamped field. I also removed a few unused update functions, as well as the circular buffer of timestamps in the `Field` class as they were unused and inconsistent with other world datatypes.

I'm open to alternative class names since the `Field` is slightly different from other structures like `Ball`, where there's a `BallState`, `TimestampedBallState`, and then a `Ball` class that stores a list of timestamped states. The `Field` has no equivalent for storing a list of states. It may make more sense to just have `FieldState` and `TimestampedFieldState`. Open to suggestions.

This PR is very long because so many functions got renamed / refactored. If you focus on anything, please focus on the new `FieldState` and `Field` classes, and their corresponding unit tests. That's the core of this change.

### Testing Done

split existing unit tests into the new classes

### Resolved Issues

resolves #1384 

### Length Justification

This is 99% renaming and a few tweaks to use new state functions / constructors. Sorry it's this long, but not much I can do about it (wasn't very fun to do).

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [x] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [x] **Remove all commented out code**
- [x] **Remove extra print statements**: for example, those just used for testing
- [x] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [x] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

**Review by jonathanlew — CHANGES_REQUESTED** [review]
I'd suggest `TimestampedField` composed of `Field` and `Timestamp`. Good to see that we're getting rid of the unused CircularBuffer

**Comment by MathewMacDougall** [comment]
> I'd suggest `TimestampedField` composed of `Field` and `Timestamp`. Good to see that we're getting rid of the unused CircularBuffer

done

**Review by jonathanlew — CHANGES_REQUESTED** [review]
Sorry change of plan: **_yeet_** `Timestamp`s from `Field`, since we don't use it at all: 
1. Check out my branch: https://github.com/jonathanlew/Software/tree/remove_timestamp_from_field 
-> `Timestamp` was only used for tests and in `World`
2. The `Timestamp` that `World` gets from `Field` is always 0. Look at network_filter.cpp: 
```
    Field field =
        Field(field_length, field_width, defense_length, defense_width, goal_width,
              boundary_width, center_circle_radius, Timestamp::fromSeconds(0));
```
^ We don't even update a new `Field` with an up to date `Timestamp`. Since we don't publish `World` if `Field` is invalid (#1218), the `Timestamp` in `Field` is irrelevant, anyways.

So, please :pray: also yeet :rocket: `Timestamp::fromSeconds(0)` from all the `Field` constructors and delete all the tests that use/update `Timestamp`.

**Comment by MathewMacDougall** [comment]
> Sorry change of plan: yeet Timestamps from Field, since we don't use it at all:

Good catch. Timestamps have been :rocket: **_YEETED_** :rocket: . Wish I did this from the start, this was a lot less work :cry:

File: src/software/world/world.cpp [hunk 1]
```diff
@@ -159,7 +158,7 @@ Timestamp World::getMostRecentTimestampFromMembers()
     // Add all member timestamps to a list
     std::initializer_list<Timestamp> member_timestamps = {
         friendly_team_.getMostRecentTimestamp(), enemy_team_.getMostRecentTimestamp(),
-        ball_.getPreviousStates().front().timestamp(), field_.getMostRecentTimestamp()};
+        ball_.getPreviousStates().front().timestamp()};
```
Comment (jonathanlew) [inline src/software/world/world.cpp:161]: Would be good to update the comment for `getMostRecentTimestampFromMembers` in world.h to something like

`     * Searches all timestamped member objects of world for the most recent Timestamp value`

**Review by jonathanlew — COMMENTED** [review]
Excellent yeetage :rocket: :tada:

**Comment by jonathanlew** [comment]
Can we change the title of the PR?

**Review by akhilveeraghanta — DISMISSED** [review]
lgtm, pending merge conflict

**Review by garethellis0 — DISMISSED** [review]
Well the "yeetage" here certainly made it easy to review. LGTM :+1: :tada:

File: src/software/world/world.cpp (hunk 1 above)
Comment (MathewMacDougall) [reply src/software/world/world.cpp:161]: good idea. updated comment as suggested

**Comment by MathewMacDougall** [comment]
Fixed merge conflicts and updated javadoc comment. Ready for re-review.

**Review by jonathanlew — DISMISSED** [review]
:fire:

File: src/software/world/field.h [hunk 2]
```diff
@@ -35,6 +35,12 @@ typedef enum
 class Field
 {
    public:
+    /**
+     * Default constructs a field with all dimensions set to 0. This default
+     * field will be invalid (ie. isValid() returns false)
```
Comment (EvanMorcom) [inline src/software/world/field.h:40]: What was the idea behind this default constructor?

File: src/software/world/field.h [hunk 3]
```diff
@@ -353,21 +332,21 @@ class Field
      *
      * @return the size of the margin/bounds around the field
```
Comment (EvanMorcom) [inline src/software/world/field.h]: Maybe add some units here

**Review by EvanMorcom — COMMENTED** [review]
LGTM 🍞
Just a couple minor comments.
Get that 🅱️ read

File: src/software/world/field.h (hunk 2 above)
Comment (MathewMacDougall) [reply src/software/world/field.h:40]: We have `isValid` again so have the ability to check for default-constructed / invalid fields. And we were _constantly_ constructing fields as `Field(0, 0, 0, 0, 0, 0, 0)` so I decided a default constructor was worth it again.

File: src/software/world/field.h (hunk 3 above)
Comment (MathewMacDougall) [reply src/software/world/field.h]: added

**Review by jonathanlew — APPROVED** [review]

**Review by garethellis0 — APPROVED** [review]

**Review by LiCody — APPROVED** [review]

**Pull request merged**